### ISSUE TITLE
DataStore doesn't accept fields starting with underscore

### ISSUE THREAD

**Issue by amercader** [ckan/ckan#6446]
*TL;DR*
* DataStore currently doesn't allow fields prefixed with underscore. In general we should allow this.
* There are some internal fields added by the DataStore which are underscore-prefixed, which can clash with incoming user-provided fields, most notably `_id`. Although the best approach would be to rename this field to something else, this field has been returned as part of the API since the first version of the DataStore, so this could break existing integrations.
---
The current implementation of the Postgres DataStore backend (ie the default one use by most users) doesn't allow to submit tables with fields that start with underscores in it.

For instance:
```
curl -X POST http://localhost:5010/api/action/datastore_create -H "Authorization:`cat ../api_token_amercader.txt`" -d '{"resource": {"package_id":"test-datapusher"},"fields": [{"id": "_a", "type": "numeric"},{"id": "_b", "type":"numeric"}],"records": [{"_a":1, "_b":1},{"_a": 2, "_b": 2}]}'
```

Gets you:
```
{
    "help": "http://localhost:5010/api/3/action/help_show?name=datastore_create",
    "error": {
        "fields": [
            "\"_a\" is not a valid field name"
        ],
        "__type": "Validation Error"
    },
    "success": false
}
```
Omitting the field definitions also fails because the underscore fields are not picked up:
```
curl -X POST http://localhost:5010/api/action/datastore_create -H "Authorization:`cat ../api_token_amercader.txt`" -d '{"resource": {"package_id":"test-datapusher"},"records": [{"_a":1, "_b":1},{"_a": 2, "_b": 2}]}'

{
    "help": "http://localhost:5010/api/3/action/help_show?name=datastore_create",
    "error": {
        "fields": [
            "\"_a\" is not a valid field name"
        ],
        "__type": "Validation Error"
    },
    "success": false
}
```

A way to address this could be modifying the incoming field names if they start with underscore like in #6421 but this doesn't feel right from an integration point of view, and might conflict with defined schemas, existing tools, etc. 

We should instead consider removing the limitation. Postgres of course has no issue with underscore-prefixed columns, so this limitation seems arbitrary.  It seems like it would be straightforward to remove [existing](https://github.com/ckan/ckan/blob/de31140c4860aacf38c10aeed379751f75cc012e/ckanext/datastore/backend/postgres.py#L294) limitations and allow these fields but reading the code it looks like underscore fields might be used to flag [array columns](https://github.com/ckan/ckan/blob/de31140c4860aacf38c10aeed379751f75cc012e/ckanext/datastore/backend/postgres.py#L347] (?)
This sounds like internal implementation details that could potentially be refactored.

But then we have the issue of `_id`. In this particular case, this internal field is added automatically to all tables. Renaming it to something like `__ckan_id` to avoid clashes would be ideal but this might be a disruptive change. 

It is used internally [for sorting](https://github.com/ckan/ckan/blob/de31140c4860aacf38c10aeed379751f75cc012e/ckanext/datastore/backend/postgres.py#L1327), which could be refactored if the field was renamed but in addition this field has been returned as part of the API since the very first DataStore version (ie a long time) so there might be external tools and clients relying on its presence. For instance the [datatables](https://github.com/ckan/ckan/blob/master/ckanext/datatablesview/public/datatablesview.js#L309) view does, recline might too and likely others.

Is it worth keeping this field unchanged (and rename any user-provided `_id` fields)?
Shall we rename it and document it?

**Comment by jqnatividad** [comment]
I suggest we at least document it.

I found this out the hard way with Datapusher+.

DP+ "sanitizes" header names to ensure that the column names used to create the postgres tables are valid names, with a default to [steer users away even from valid quoted identifiers](https://lerner.co.il/2013/11/30/quoting-postgresql/).

https://github.com/jqnatividad/qsv/blob/master/src/cmd/safenames.rs#L2
https://github.com/jqnatividad/qsv/blob/master/tests/test_safenames.rs

With the rules largely based on Postgres valid identifier definition
https://www.postgresql.org/docs/12/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS



To ensure that the resulting column names are still readable, DP+ saves the original column name in the Data Dictionary Label field, which DataTables view uses to label the column instead of the underlying sanitized postgresql name.
